src/EventHandler.php

```diff
@@ -27,69 +27,200 @@ class EventHandler
-            }
-
-            $summary = new ResultSummary(
-                new SummaryCounters(),
-                new DatabaseInfo('n/a'),
-                new CypherList([]),
-                null,
-                null,
-                $statement,
-                QueryTypeEnum::READ_ONLY(),
-                0,
-                0,
-                new ServerInfo(Uri::create(''), ConnectionProtocol::BOLT_V5(), 'n/a'),
+            $this->stopwatch?->stop($stopwatchName);
+            $event = new FailureEvent(
+                alias: $alias,
+                statement: $statement,
+                exception: $e,
+                time: new \DateTimeImmutable('now'),
+                scheme: $scheme,
+                transactionId: $transactionId
             );
 
-            $tbr = new SummarizedResult($summary);
+            $this->dispatcher?->dispatch($event, FailureEvent::EVENT_ID);
+
+            throw $e;
         }
+    }
+
+    /**
+     * @return array{'preEvent': object, 'preEventId': string, 'postEvent': object, 'postEventId': string}
+     */
+    private function createPreAndPostEventsAndIds(
+        TransactionState $nextTransactionState,
+        string $alias,
+        string $scheme,
+        string $transactionId,
+    ): array {
+        [$preEvent, $preEventId] = match ($nextTransactionState) {
+            TransactionState::ACTIVE => [
+                new PreTransactionBeginEvent(
+                    alias: $alias,
+                    time: new \DateTimeImmutable(),
+                    scheme: $scheme,
+                    transactionId: $transactionId,
+                ),
+                PreTransactionBeginEvent::EVENT_ID,
+            ],
+            TransactionState::TERMINATED, TransactionState::ROLLED_BACK => [
+                new PreTransactionRollbackEvent(
+                    alias: $alias,
+                    time: new \DateTimeImmutable(),
+                    scheme: $scheme,
+                    transactionId: $transactionId,
+                ),
+                PreTransactionRollbackEvent::EVENT_ID,
+            ],
+            TransactionState::COMMITTED => [
+                new PreTransactionCommitEvent(
+                    alias: $alias,
+                    time: new \DateTimeImmutable(),
+                    scheme: $scheme,
+                    transactionId: $transactionId,
+                ),
+                PreTransactionCommitEvent::EVENT_ID,
+            ],
+        };
+        [$postEvent, $postEventId] = match ($nextTransactionState) {
+            TransactionState::ACTIVE => [
+                new PostTransactionBeginEvent(
+                    alias: $alias,
+                    time: new \DateTimeImmutable(),
+                    scheme: $scheme,
+                    transactionId: $transactionId,
+                ),
+                PostTransactionBeginEvent::EVENT_ID,
+            ],
+            TransactionState::TERMINATED, TransactionState::ROLLED_BACK => [
```

Comment:
Terminated only happens in case of a Neo4j Exception, which we handle in the Failure Event. In theory, Terminated never happens here so we should just match it and throw an exception

Reply:
UnexpectedValueException is a good candidate